Improve substepping implementation in InelasticDefgradTransvIsotropElastViscoplast

dragos-ana opened the pull request:
<!--
* Title: Provide a general summary of your changes in the Title above.

* Assignees:  If you know anyone who should likely handle bringing this pull request (PR) to completion -- that's usually yourself -- select them from the Assignees drop-down on the right.

* Labels: Update the label of the issue(s) addressed by this pull request to "Under Review".
-->

## Description and Context
<!--
Provide a brief and concise description of your proposed change. Questions you should think about:
* Why is this change required?  What problem does it solve?
* Is there a bigger picture? Is this PR a part of a larger set of changes? Which further steps are planned after merging this PR, if any?
* How has the proposed implementation been verified and tested?

Keep the description of the PR always up-to-date and concise.
-->
The implementation of substepping in the time integration of InelasticDefgradTransvIsotropElastViscoplast currently uses try-catch mechanisms, with explicit throwing of exceptions to trigger a new substep. As discussed in [2646](https://gitlab.lrz.de/baci/baci/-/merge_requests/2646#note_3454413), it is desirable to avoid this type of mechanism. Moreover, simulations using the current implementation have shown various convergence problems and "random" overflow errors. This PR addresses the issues outlined herein. Specifically, the following work was performed:

- Categorized the types of errors that can be encountered in the time integration, and in the additional stiffness calculation (contribution to the global linearization).
- Removed the try-catch handling and improved the substepping management by using an error status integer indicating which of the possible errors were encountered. Thus, substepping is now triggered using this error status. This also applies for the perturbation-based linearization, which is triggered if an error is encountered in the analytical linearization. The perturbation-based linearization can now be found in a separate function `evaluate_additional_cmat_perturb_based`.
- The overflow error checking is now also performed for the logarithmic substepping, but with other numerical bounds since the computation of the update tensor is not required. The "magic" small value of the time step used previously by default for logarithmic substepping (1.0e-100) was removed.
- Also added overflow error checking for the derivative evaluation of the implemented Reformulated Johnson-Cook law, by specifically calculating the logarithm of both derivatives and checking their numerical values against set numerical bounds. This, together with the previous point, solved the problem of "random" overflow errors.
- Several other small changes to improve the code readability, such as the addition of the dedicated `check_predictor` method or the `SubstepParams` struct containing substepping params. 



## Related Issues and Merge Requests
<!--
If applicable, let us know how this pull request is related to any other open issues or pull requests by linking to them here.
Some suggestion for keywords:
Closes (will automatically close mentioned issue if merged), Blocks, Related to
-->
[2646](https://gitlab.lrz.de/baci/baci/-/merge_requests/2646#note_3454413)

c-p-schmidt commented on src/mat/4C_mat_inelastic_defgrad_factors.cpp:
```diff
@@ -26,13 +26,61 @@
 {
+  // enum class for error types in InelasticDefgradTransvIsotropElastViscoplast, used for triggering
+  // the substepping procedures
+  enum class ErrorType
```
One question on the error handling. In other methods, you use an `int` to encode the error. Wouldn't it be nicer and more descriptive to also make use of such an enum?

dragos-ana replied:
As explained below, I use an int to signal an error in the viscoplastic law evaluations anyway, to avoid providing the ErrorType enum class to it as well.  Since the conversion from `int` to `ErrorType` is then performed anyway,  I would see this as a matter of taste whether we pass ErrorTypes or ints to and from the material methods.

sebproell commented on src/mat/4C_mat_inelastic_defgrad_factors.cpp:
```diff
@@ -2878,164 +2850,147 @@ Core::LinAlg::Matrix<10, 1> Mat::InelasticDefgradTransvIsotropElastViscoplast::l
-  unsigned int max_num_of_substeps = 1;
+  // initialize substep parameters
+  SubstepParams substep_params = {
+      0,  // t = 0 (time parameter)
+      1,  // substep_counter = 1 (evaluation of first substep)
+      time_step_settings_
+          .dt_,  // curr_dt = time_step_settings_.dt_ (first substep length = full time step)
+      0,         // time_step_halving_counter = 0 (full time step, therefore no substep halving yet)
+      1,         // max_num_of_substeps = 1 (1 substep to evaluate: full time step)
+      0,         // iter = 0 (0 LNL iterations for the current substep)
```
You can use designated initializers here: 

```suggestion
      .t = 0, 
     .substep_counter =  1, 
     ...
```

sebproell commented on src/mat/vplast/4C_mat_vplast_reform_johnsoncook.cpp:
```diff
@@ -137,13 +146,26 @@ Mat::Viscoplastic::ReformulatedJohnsonCook::evaluate_derivatives_of_plastic_stra
+    {
+      err_status = 2;
+      return Core::LinAlg::Matrix<2, 1>{true};
+    }
```
You are computing the same expensive functions, `log` and `exp`, and sometimes longer subexpressiones like `std::log(p) + e * (equiv_stress / yield_strength - 1.0) + std::log(e)` multiple times. This results in way more instructions than you have to do and you could extract the subexpressions.

sebproell replied:
Small example where I rearranged the computations to make this specific part 2.5 times faster(I did not even include the overflow check): https://quick-bench.com/q/ELEz4rotYR56n1y1IThkiFlZxWU

You don't need to do that in this PR. I just wanted to point out that this code looks really expensive and since you do this on the quadrature point I can imagine these calculations contribute significantly to your total run time.

dragos-ana replied:
I have included your suggestions now. Thanks a lot for the hint and for the offline discussion!

dragos-ana commented on src/mat/4C_mat_inelastic_defgrad_factors.hpp:
```diff
@@ -1565,10 +1620,52 @@ namespace Mat
+     * @param[out] curr_CM current right Cauchy-Green deformation tensor, interpolated using
+     * the reference matrices of the time step (interpolated again within this method with the
+     * updated new substep length)
+     * @return error status for the new substep (0: no errors, 1: we have halved the time step too
```
No, I only expect the two states. Thus, I now use a boolean as proposed, thanks!

dragos-ana commented on src/mat/4C_mat_inelastic_defgrad_factors.hpp:
```diff
@@ -1502,6 +1507,35 @@ namespace Mat
+      double curr_dt;
+      //! number of times the problem time step \f$ \Delta t \f$ has been halved
+      unsigned int time_step_halving_counter;
+      //!  current maximum number of substeps to be evaluated within the time step \f$ \Delta t
```
Yes, makes more sense. Thanks!

dragos-ana commented on src/mat/4C_mat_inelastic_defgrad_factors.cpp:
```diff
@@ -26,13 +26,61 @@
+  // the substepping procedures
+  enum class ErrorType
+  {
+    NegativePlasticStrain = 1,  // negative plastic strain which does not allow for evaluations
```
This is helpful since I need to "translate" from an int to an ErrorType when communicating with the viscoplastic law. E.g. when returning the error status from `evaluate_plastic_strain_rate` in  `4C_mat_vplast_reform_johnsoncook.cpp`, I want to convert this int to an error type inside InelasticDefgradTransvIsotropElastViscoplast without providing the ErrorType class also to the viscoplasticity law.

dragos-ana commented on src/mat/4C_mat_inelastic_defgrad_factors.cpp:
```diff
@@ -2300,164 +2358,97 @@ void Mat::InelasticDefgradTransvIsotropElastViscoplast::evaluate_additional_cmat
-  }
+
+  // declare error status (0: no errors)
+  int err_status = 0;
```
This is the error status corresponding to the ErrorType. But as explained in both threads above, a conversion int <-> ErrorType has to be performed anyway in the current implementation of error handling.

dragos-ana commented:
Thank you for the detailed review, @c-p-schmidt  and @sebproell!

dragos-ana commented:
Based on the discussions above, I removed the `ErrorType<->int` conversion logic entirely by creating the header file `4C_mat_inelastic_defgrad_factors_service.hpp`, which I provide to both `InelasticDefgradTransvIsotropElastViscoplast` and `Viscoplastic::Law`. Additionally, I implemented some small optimizations, such as replacing division by multiplication where it was possible and reasonable to do so.

sebproell commented on src/mat/4C_mat_inelastic_defgrad_factors_service.hpp:
```diff
@@ -0,0 +1,49 @@
+                                   // failed
+  };
+
+  /// map: error types to error messages in InelasticDefgradTransvIsotropElastViscoplast
+  extern std::map<Mat::ViscoplastErrorType, std::string> ViscoplastErrorMessages;
```
Should this simply be a function `to_string(ViscoPlastErrorType)`?

sebproell reviewed (approved):
Only a minor comment for future improvement.
